azure-iot-device/azure/iot/device/common/pipeline/operation_flow.py

```diff
@@ -0,0 +1,243 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
```

Comment:
Curious why this became a separate module, given that the first param of ever function contained within is `stage`. Is it just for ease of testing? I'd be a bit concerned if we had to restructure code for test purposes.

Reply:
I pulled these out because they're helper functions that can be used by a stage, but they're not necessarily part of the identity of a stage.  If I kept them inside the PipelineStage object, they would be more appropriately treated as private functions inside the stage.  But, these helpers provide more functionality than any one stage might need.  In that way, keeping these as privates inside PipelineStage would make each stage much more powerful than it needs to be.

Another way to think about it is to ask if every method on PipelineStage (or any stage) has to do with the core functionality of that stage.  A stage can run operations and handle events.  If we take the first helper (run_ops_in_serial), I would argue that running operations in serial is not part of the identity of a stage.  Rather, running ops in serial is more of an external helper function.

---
In reply to: [293418167](https://github.com/Azure/azure-iot-sdk-python-preview/pull/106#discussion_r293418167) [](ancestors = 293418167)